File: src/main/java/com/akagiyui/drive/model/ResponseResult.java
package com.akagiyui.drive.model;

import com.akagiyui.drive.component.ResponseEnum;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletResponse;
import lombok.Data;
import org.springframework.http.HttpStatus;

/**
 * 响应包装体
 *
 * @author AkagiYui
 */
@Data
public class ResponseResult<T> {
    /**
     * 状态码
     */
    private Integer code;
    /**
     * 消息
     */
    private String msg;
    /**
     * 数据
     */
    private T data;

    /**
     * 通用响应
     *
     * @param code 状态码
     * @param msg  消息
     * @param data 数据
     * @param <T>  数据类型
     * @return 响应体
     */
    public static <T> ResponseResult<T> response(Integer code, String msg, T data) {
        ResponseResult<T> result = new ResponseResult<>();
        result.setCode(code);
        result.setMsg(msg);
        result.setData(data);
        return result;
    }

    /**
     * 通用响应
     *
     * @param status 状态枚举
     * @return 响应体
     */
    public static ResponseResult<?> response(ResponseEnum status) {
        return response(status.getCode(), status.getMsg(), null);
    }

    /**
     * 通用响应
     *
     * @param status 状态枚举
     * @return 响应体
     */
    public static <T> ResponseResult<T> response(ResponseEnum status, T data) {
        return response(status.getCode(), status.getMsg(), data);
    }

    /**
     * 成功响应
     *
     * @param data 数据
     * @return 响应体
     */
    public static <T> ResponseResult<T> success(T data) {
        ResponseEnum success = ResponseEnum.SUCCESS;
        return response(success.getCode(), success.getMsg(), data);
    }

    /**
     * 无数据成功响应
     *
     * @return 响应体
     */
    public static ResponseResult<?> success() {
        return success(null);
    }

    /**
     * 把响应内容写入响应体
     *
     * @param res    响应
     * @param code   状态码
     * @param status 状态枚举
     */
    public static void writeResponse(HttpServletResponse res, HttpStatus code, ResponseEnum status) {
        res.setStatus(code.value());
        res.setContentType("application/json;charset=UTF-8");
        res.setCharacterEncoding("UTF-8");
        try {
            ObjectMapper objectMapper = new ObjectMapper();
            res.getWriter().write(objectMapper.writeValueAsString(response(status)));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * 把响应内容写入响应体
     *
     * @param res    响应对象
     * @param code   状态码
     * @param object 响应内容
     */
    public static void writeResponse(HttpServletResponse res, HttpStatus code, Object object) {
        res.setStatus(code.value());
        res.setContentType("application/json;charset=UTF-8");
        res.setCharacterEncoding("UTF-8");
        try {
            ObjectMapper objectMapper = new ObjectMapper();
            res.getWriter().write(objectMapper.writeValueAsString(response(ResponseEnum.SUCCESS, object)));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/com/akagiyui/common/ResponseResult.kt
package com.akagiyui.common

import com.akagiyui.common.delegate.LoggerDelegate
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpStatus

/**
 * 响应包装体
 *
 * @param T 数据类型
 * @property code 状态码
 * @property msg 消息
 * @property data 数据
 */
data class ResponseResult<T>(
    val code: Int,
    val msg: String,
    val data: T? = null,
) {
    companion object {
        private val log by LoggerDelegate()

        /**
         * 通用响应
         */

        fun <T> response(code: Int, msg: String, data: T? = null): ResponseResult<T> {
            return ResponseResult(code, msg, data)
        }

        /**
         * 通用响应
         */
        fun <T> response(status: ResponseEnum): ResponseResult<T> {
            return response(status.code, status.msg, null)
        }

        /**
         * 通用响应
         */
        fun <T> response(status: ResponseEnum, data: T?): ResponseResult<T> {
            return if (status == ResponseEnum.BAD_REQUEST && data is String && data.isNotBlank()) {
                response(status.code, data, null)
            } else {
                response(status.code, status.msg, data)
            }
        }

        /**
         * 成功响应
         */
        fun <T> success(data: T? = null): ResponseResult<T> {
            return response(ResponseEnum.SUCCESS, data)
        }

        /**
         * 把响应内容写入响应体
         */
        fun writeResponse(res: HttpServletResponse, code: HttpStatus, status: ResponseEnum) {
            res.status = code.value()
            res.contentType = "application/json;charset=UTF-8"
            res.characterEncoding = "UTF-8"
            try {
                val objectMapper = jacksonObjectMapper()
                res.writer.write(objectMapper.writeValueAsString(response(status, null)))
            } catch (e: Exception) {
                log.error("Write response error", e)
            }
        }

        /**
         * 把响应内容写入响应体
         */
        fun writeResponse(res: HttpServletResponse, code: HttpStatus, obj: Any) {
            res.status = code.value()
            res.contentType = "application/json;charset=UTF-8"
            res.characterEncoding = "UTF-8"
            try {
                val objectMapper = jacksonObjectMapper()
                res.writer.write(objectMapper.writeValueAsString(response(ResponseEnum.SUCCESS, obj)))
            } catch (e: Exception) {
                log.error("Write response error", e)
            }
        }
    }
}
